File: formats/pdf/new/c++/filters/src/predictor.cpp
#include "predictor.hpp"

#include <cstdlib>
#include <limits>

namespace {

bool checkedAdd(size_t x, size_t y, size_t &result) {
  if (y > std::numeric_limits<size_t>::max() - x) {
    return false;
  }
  result = x + y;
  return true;
}

bool checkedMultiply(size_t x, size_t y, size_t &result) {
  if (x != 0 && y > std::numeric_limits<size_t>::max() / x) {
    return false;
  }
  result = x * y;
  return true;
}

uint8_t paethPredictor(uint8_t left, uint8_t above, uint8_t upperLeft) {
  int const estimate = int(left) + int(above) - int(upperLeft);
  int const leftDistance = std::abs(estimate - int(left));
  int const aboveDistance = std::abs(estimate - int(above));
  int const upperLeftDistance = std::abs(estimate - int(upperLeft));

  if (leftDistance <= aboveDistance && leftDistance <= upperLeftDistance) {
    return left;
  }
  if (aboveDistance <= upperLeftDistance) {
    return above;
  }
  return upperLeft;
}

uint8_t pngPrediction(
  uint8_t tag,
  uint8_t left,
  uint8_t above,
  uint8_t upperLeft
) {
  switch (tag) {
    case 0: return 0;
    case 1: return left;
    case 2: return above;
    case 3: return uint8_t((uint16_t(left) + uint16_t(above)) / 2);
    case 4: return paethPredictor(left, above, upperLeft);
    default: return 0;
  }
}

bool pngPredictor(
  uint64_t colorsValue,
  uint64_t bitsPerComponentValue,
  uint64_t columnsValue,
  std::string &bytes
) {
  if (colorsValue == 0 || columnsValue == 0) {
    return false;
  }
  if (bitsPerComponentValue != 1 &&
      bitsPerComponentValue != 2 &&
      bitsPerComponentValue != 4 &&
      bitsPerComponentValue != 8 &&
      bitsPerComponentValue != 16) {
    return false;
  }
  if (colorsValue > std::numeric_limits<size_t>::max() ||
      bitsPerComponentValue > std::numeric_limits<size_t>::max() ||
      columnsValue > std::numeric_limits<size_t>::max()) {
    return false;
  }

  size_t const colors = size_t(colorsValue);
  size_t const bitsPerComponent = size_t(bitsPerComponentValue);
  size_t const columns = size_t(columnsValue);

  size_t bitsPerPixel;
  size_t roundedBitsPerPixel;
  size_t rowBits;
  size_t roundedRowBits;
  size_t rowSize;
  if (!checkedMultiply(colors, bitsPerComponent, bitsPerPixel) ||
      !checkedAdd(bitsPerPixel, 7, roundedBitsPerPixel) ||
      !checkedMultiply(bitsPerPixel, columns, rowBits) ||
      !checkedAdd(rowBits, 7, roundedRowBits)) {
    return false;
  }

  size_t const bytesPerPixel = roundedBitsPerPixel / 8;
  size_t const rowBytes = roundedRowBits / 8;
  if (!checkedAdd(rowBytes, 1, rowSize) || bytes.size() % rowSize != 0) {
    return false;
  }
  if (bytes.empty()) {
    return true;
  }

  auto *data = reinterpret_cast<uint8_t *>(bytes.data());

  uint8_t const firstTag = data[0];
  if (firstTag > 4) {
    return false;
  }
  for (size_t column = 0; column < rowBytes; column++) {
    uint8_t const byte = data[column + 1];
    uint8_t const left =
      column < bytesPerPixel ? 0 : data[column - bytesPerPixel];
    data[column] = byte + pngPrediction(firstTag, left, 0, 0);
  }

  size_t readRow = rowSize;
  size_t write = rowBytes;
  while (readRow < bytes.size()) {
    uint8_t const tag = data[readRow];
    if (tag > 4) {
      return false;
    }

    size_t const rowEnd = readRow + rowSize;
    for (size_t read = readRow + 1, column = 0;
         read < rowEnd;
         read++, column++) {
      uint8_t const byte = data[read];
      uint8_t const left =
        column < bytesPerPixel ? 0 : data[write + column - bytesPerPixel];
      uint8_t const above = data[write - rowBytes + column];
      uint8_t const upperLeft =
        column < bytesPerPixel
          ? 0
          : data[write - rowBytes + column - bytesPerPixel];
      data[write + column] =
        byte + pngPrediction(tag, left, above, upperLeft);
    }
    write += rowBytes;
    readRow = rowEnd;
  }

  bytes.resize(write);
  return true;
}

}

bool unpredict(
  uint64_t predictor,
  uint64_t colors,
  uint64_t bpc,
  uint64_t columns,
  std::string &bytes
) {
  switch (predictor) {
    case 1: return true; // no predictor
    case 10:
    case 11:
    case 12:
    case 13:
    case 14:
    case 15:
      return pngPredictor(colors, bpc, columns, bytes);
    default: return false; // unsupported
  }
}

File: formats/pdf/new/rust/pdfcos/src/filters/predictor.rs
//! Predictor reversal for PDF LZW and Flate stream filters.

pub(crate) fn apply_predictor(
    input: Vec<u8>,
    predictor: u64,
    colors: usize,
    bits_per_component: usize,
    columns: usize,
) -> Result<Vec<u8>, String> {
    match predictor {
        1 => Ok(input),
        10..=15 => png_predictor(input, colors, bits_per_component, columns),
        _ => Err(format!(
            "unsupported predictor parameters: Predictor={predictor}, Colors={colors}, \
             BitsPerComponent={bits_per_component}"
        )),
    }
}

/// Reverse the PNG row predictors used by PDF Predictor values 10 through 15.
/// Each encoded row begins with a tag selecting None, Sub, Up, Average, or
/// Paeth. The tag is removed while the reconstructed rows are compacted in
/// place.
///
/// See ISO 32000-2:2017 section 7.4.4.4 and
/// https://en.wikipedia.org/wiki/PNG#Filtering.
fn png_predictor(
    mut input: Vec<u8>,
    colors: usize,
    bits_per_component: usize,
    columns: usize,
) -> Result<Vec<u8>, String> {
    if colors == 0 {
        return Err("invalid PNG predictor parameter Colors".to_owned());
    }
    if !matches!(bits_per_component, 1 | 2 | 4 | 8 | 16) {
        return Err("invalid PNG predictor parameter BitsPerComponent".to_owned());
    }
    if columns == 0 {
        return Err("invalid PNG predictor parameter Columns".to_owned());
    }

    let bits_per_pixel = colors
        .checked_mul(bits_per_component)
        .ok_or_else(|| "PNG predictor pixel size does not fit in usize".to_owned())?;
    let bytes_per_pixel = bits_per_pixel
        .checked_add(7)
        .ok_or_else(|| "PNG predictor pixel size does not fit in usize".to_owned())?
        / 8;
    let row_bytes = bits_per_pixel
        .checked_mul(columns)
        .and_then(|bits| bits.checked_add(7))
        .ok_or_else(|| "PNG predictor row size does not fit in usize".to_owned())?
        / 8;
    let row_size = row_bytes
        .checked_add(1)
        .ok_or_else(|| "PNG predictor row size does not fit in usize".to_owned())?;
    if input.len() % row_size != 0 {
        return Err("truncated PNG predictor row".to_owned());
    }
    if input.is_empty() {
        return Ok(input);
    }

    let first_tag = input[0];
    validate_png_predictor_tag(first_tag)?;
    for column in 0..row_bytes {
        let byte = input[column + 1];
        let left = if column < bytes_per_pixel {
            0
        } else {
            input[column - bytes_per_pixel]
        };
        input[column] = byte.wrapping_add(png_prediction(first_tag, left, 0, 0));
    }

    let mut read_row = row_size;
    let mut write = row_bytes;
    while read_row < input.len() {
        let tag = input[read_row];
        validate_png_predictor_tag(tag)?;

        let row_end = read_row + row_size;
        for (column, read) in (read_row + 1..row_end).enumerate() {
            let byte = input[read];
            let left = if column < bytes_per_pixel {
                0
            } else {
                input[write + column - bytes_per_pixel]
            };
            let above = input[write - row_bytes + column];
            let upper_left = if column < bytes_per_pixel {
                0
            } else {
                input[write - row_bytes + column - bytes_per_pixel]
            };
            input[write + column] =
                byte.wrapping_add(png_prediction(tag, left, above, upper_left));
        }
        write += row_bytes;
        read_row = row_end;
    }

    input.truncate(write);
    Ok(input)
}

fn validate_png_predictor_tag(tag: u8) -> Result<(), String> {
    if tag <= 4 {
        Ok(())
    } else {
        Err(format!("unsupported PNG predictor tag {tag}"))
    }
}

fn png_prediction(tag: u8, left: u8, above: u8, upper_left: u8) -> u8 {
    match tag {
        0 => 0,
        1 => left,
        2 => above,
        3 => ((u16::from(left) + u16::from(above)) / 2) as u8,
        4 => paeth_predictor(left, above, upper_left),
        _ => unreachable!("PNG predictor tag was validated"),
    }
}

fn paeth_predictor(left: u8, above: u8, upper_left: u8) -> u8 {
    let left = i32::from(left);
    let above = i32::from(above);
    let upper_left = i32::from(upper_left);
    let estimate = left + above - upper_left;
    let left_distance = (estimate - left).abs();
    let above_distance = (estimate - above).abs();
    let upper_left_distance = (estimate - upper_left).abs();

    if left_distance <= above_distance && left_distance <= upper_left_distance {
        left as u8
    } else if above_distance <= upper_left_distance {
        above as u8
    } else {
        upper_left as u8
    }
}
